pandas/io/pytables.py

```diff
@@ -5076,7 +5076,7 @@ def _dtype_to_kind(dtype_str: str) -> str:
     return kind
 
 
-def _get_data_and_dtype_name(data: ArrayLike):
+def _get_data_and_dtype_name(data: AnyArrayLikeUnion):
```

Comment:
might be better to ensure we unwrap to avoid Index/Series?

for my edification: i was under the impression that AnyArrayLike vs AnyArrayLikeUnion only mattered when we had more than one of them in a signature; i.e. 

```
def foo(arg: AnyArrayLike) -> AnyArrayLike:
```

implies the return type matches the arg type, while using `AnyArrayLikeUnion` does not imply that.  if that's correct, does it matter which we use in cases like this with only one annotation in the signature?

Reply:
in this case we change the type of data in the body. i.e assign to data and hence no preserving type of data.

I've used ..Union to be consistent with FrameOrSeries.

we could change ArrayLike and AnyArrayLike to be unions and ArrayLikeT and AnyArrayLikeT for type variables.

or we could assign to another variable name in the body (simpler, but imo messy and less clear)

Reply:
im happy to follow your lead on this

Reply:
@WillAyd wdyt?

Reply:
> or we could assign to another variable name in the body (simpler, but imo messy and less clear)

I don't think worth going down this path. We will likely need a Union type for variable type annotations anyway, since if we use ArrayLike we would get `error: Type variable "pandas._typing.ArrayLike" is unbound  [valid-type]`